### PR TITLE
[PkgServer] use Tar.jl to extract tarballs

### PR DESCRIPTION
**Pull request by johnnychen94**
cont. #29

### CODE REVIEW

File: src/resource.jl [hunk 1]
```diff
@@ -21,6 +21,9 @@ const resource_re = Regex("""
 """, "x")
 const hash_part_re = Regex("/($hash_re)\$")
 
+compress(io::IO) = TranscodingStream(GzipCompressor(level=9), io)
```
Comment (GunnarFarneback) [inline src/resource.jl]: This is unused.

File: src/resource.jl [hunk 2]
```diff
@@ -21,6 +21,9 @@ const resource_re = Regex("""
...
+decompress(io::IO) = TranscodingStream(GzipDecompressor(), io)
```
Comment (GunnarFarneback) [inline src/resource.jl]: No need to involve `TranscodingStreams`. `GzipDecompressorStream` can be used directly.

**Comment by johnnychen94** [comment]
Similar to https://github.com/JuliaPackaging/PkgServer.jl/pull/29/files#r417042956, there seems also a significant performance issue here.

```julia
function extract_tar_julia(tarball, outdir)
    open(tarball) do io
        Tar.extract(decompress(io), outdir)
    end
end

function extract_tar(tarball, outdir)
    mkpath(outdir)
    run(`tar -C $outdir -zxf $tarball`)
end

# tarball is a file of size 3M
@time extract_tar_julia(tarball, "tmp") # 0.06s
@time extract_tar(tarball, "tmp") # 0.03s
```

If the performance regression is accepted during building a storage server, that during serving contents via pkg server would typically harm user experiences. Perhaps this PR is not wanted at present.

File: src/resource.jl (hunk 2 above)
Comment (StefanKarpinski) [reply src/resource.jl]: Can you give an example of how that would be done? I've only used it via TranscodingStreams so far...

**Comment by StefanKarpinski** [comment]
This only happens on the first hit for a given tarball, so I'm not too worried about speed—the first person to get this is going to have a slow experience anyway. All the ones after that will be fast. We can, in principle, verify tarballs without unpacking them (they do need to be uncompressed),  but that would need to be implemented.

File: src/resource.jl (hunk 2 above)
Comment (GunnarFarneback) [reply src/resource.jl]: See https://github.com/GunnarFarneback/LocalPackageServer.jl/blob/734490f1db4e0e1597fc262566c897931781b1c0/src/resource.jl#L115.

Without checking I'm fairly certain it's exactly the same under the hood, you can just avoid a direct dependency on TranscodingStreams.

**Comment by GunnarFarneback** [comment]
> but that would need to be implemented.

https://github.com/JuliaIO/Tar.jl/pull/35

**Comment by staticfloat** [comment]
As a part of https://github.com/JuliaPackaging/PkgServer.jl/commit/3aa8ea3a6a5807d6c13c405e91d665dcfbeca3a0 I actually ended up using the within-tarball tree hash calculation that Stefan implemented in `Tar.jl`, so that tarballs never need to be extracted on-disk.  We also start calculating the tree hash while the tarball is still being downloaded, so this gives a small but important latency decrease.

Thanks for the work here @johnnychen94!

**Pull request closed without merging**